### PR TITLE
Implement filtsmooth methods as generators

### PR DESCRIPTION
**Pull request by schmidtjonathan**
This PR implements Kalman filtering methods as generators in order to make partial results accessible.

### CODE REVIEW

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:66]: If you have a more elegant solution, feel free to request changes.
I thought about something like

```python
return next(reversed(self.iterated_filtsmooth_generator(...)))
```

but I think it might be less efficient. If efficiency is relevant at all here ....

Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:66]: I am not sure about your second proposal. I like the for loop, but you dont need the `result = iter_result` assignment I think. ```for iter_result in self.iterated_filtsmooth_generator(...): pass; return iter_result``` would work, right? 

Apparently (https://stackoverflow.com/questions/2138873/cleanest-way-to-get-last-item-from-python-iterator) ```*_, last = self.iterated_filtsmooth_generator(...)``` would also be an option? I have not checked this though

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:180]: what is the idea behind this change? I am not sure how comfortable I am about changing the number of return values depending on an input argument. However, if there is a good reason to do this, I am all ears :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:66]: checked it, the `*_, last = ...` thing works :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:180]: Merely for backwards compatibility. Do you prefer having it returned in all cases?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:66]: Okay. I like the second proposal! The first, with the `for ... pass ...` loop doesn't work -- that I checked.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:180]: Hmmm. We dont do this anywhere. Maybe make this info a part of the posterior? Would that be something interesting? If you like, feel free to come by my office and we discuss this. I suppose this would have a wider impact than `Kalman` (but would also apply to the ODE solvers for instance)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:66]: Okay nevermind, your first proposal works, pylint was just telling me that it isn't a good idea. However, I would ignore pylint in this case and use the first proposal because this way, the objects aren't kept in memory.

File: docs/source/tutorials/filtsmooth/particle_filtering.ipynb
Comment (pnkraemer) [inline docs/source/tutorials/filtsmooth/particle_filtering.ipynb:2783]: I cant believe it, a readable notebook diff??

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:38]: Not complaining about the variable name, but I would've expected you go for `initial_posterior` here ;)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:53]: super minor comment: I find this a bit hard to parse. But that might just be me reading this at the end of my workday... ;)

Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:53]: Definitely feel free to ignore htis!

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:74]: when would this valueerror happen? And would the situation you describe in the message rather be a warning than an error?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:117]: super minor again: why `info_dicts` (as opposed to `info_dict`)?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:210]: naming question: is this a `filter_generator` or something along the lines of a `filter_solution_generator`?

**Review by pnkraemer — COMMENTED** [review]
Great job overall! I have only super minor comments (mostly naming questions, which -- I am aware -- is super subjective), which you may ignore if you like.

The only one that might be worth discussing is: is the filtering generator method a `filter_generator` or something along the lines of a  `filtering_solution_generator` or `filtering_posterior_generator` (`filtering_distribution_generator`?).

Once this is settled I will approve! :)

File: docs/source/tutorials/filtsmooth/particle_filtering.ipynb
Comment (schmidtjonathan) [reply docs/source/tutorials/filtsmooth/particle_filtering.ipynb:2783]: I edited the raw file 😄

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:210]: decision: `filtered_states_generator`

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:74]: Never. Decision: remove

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:118]: just checking: so you decided against the "else" solution here? :)

**Review by pnkraemer — APPROVED** [review]
See my final comment. Go ahead and merge though, you have my approval :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:118]: Nope, forgot, thanks for the reminder! Just one sec

**Pull request merged**
